src/node/proxy.ts

```diff
@@ -50,15 +50,18 @@ const maybeProxy = (req: Request): string | undefined => {
  * through to allow the redirect and login flow.
  */
 const shouldFallThrough = (req: Request): boolean => {
-  // The ideal would be to have a reliable way to detect if this is a request
-  // for (or originating from) our root or login HTML. But requests for HTML
-  // don't seem to set any content type.
-  return (
-    req.headers["content-type"] !== "application/json" &&
-    ((req.originalUrl.startsWith("/") && req.method === "GET") ||
-      (req.originalUrl.startsWith("/static") && req.method === "GET") ||
-      (req.originalUrl.startsWith("/login") && (req.method === "GET" || req.method === "POST")))
-  )
+  // See if it looks like a request for the root or login HTML.
```

Comment:
It was in response to this issue: https://github.com/cdr/code-server/issues/2128

The gist is that if we return a 40x (if we let all requests fall through) or a 30x (if we redirect all of them) then it masks the real problem (that the request isn't authenticated). 

So in the above issue the application makes xhr POST requests against the root but by default it doesn't include cookies. So at the time it fell through and got a 400 for those requests which made it look like code-server wasn't proxying correctly.

If we instead redirected all the requests to the login I imagine it'd try to parse it as JSON and would error with something like "unexpected token <".

I wasn't confident there wouldn't be more cases like this so I let through the absolute minimum required to log in.

Reply:
> If we instead redirected all the requests to the login I imagine it'd try to parse it as JSON and would error with something like "unexpected token <".

If the requester accepts html, we should redirect to login every time. If they do not, then we should return a 401 with a error message that there is no code-server cookie.

Reply:
Fair enough, I didn't do this initially because I was concerned some framework somewhere uses it in xhr requests or some noob wouldn't set the right header and blame code-server. Seems like a safe enough bet though, and it means you don't have to be at the root of your app to get redirected which is definitely nice.